app/controllers/schools_controller.rb

```diff
@@ -104,9 +106,127 @@ def update_code_of_conduct
     redirect_to standing_school_path
   end
 
+  # GET /school/discord
+  def discord_configuration
+    authorize current_school
+  end
+
+  # GET /school/discord_server_roles?cause=
+  def discord_server_roles
+    authorize current_school
+
+    if params[:cause].present? && params[:cause] == "DeletedRoles"
+      @sync_service = Discord::SyncRolesService.new(school: current_school)
+    end
+  end
+
+  # PATCH /school/discord_configuration
+  def discord_credentials
+    authorize current_school
+
+    bot_token = params["bot_token"]
+
+    discord_config = current_school.configuration.dig("discord") || {}
+    discord_config["server_id"] = params["server_id"]
+    discord_config["bot_user_id"] = params["bot_user_id"]
+    discord_config["bot_token"] = bot_token.presence ||
+      discord_config.dig("bot_token")
+
+    current_school.update!(
+      configuration:
+        current_school.configuration.merge({ "discord" => discord_config })
+    )
+
+    flash[:success] = I18n.t(
+      "schools.discord_credentials.discord_config_stored"
+    )
+
+    redirect_to discord_configuration_school_path
+  end
+
+  # POST /school/discord_sync_roles
+  def discord_sync_roles
+    authorize current_school
+
+    role_sync_service = Discord::SyncRolesService.new(school: current_school)
```

Comment:
This `Discord::SyncRolesService` seems to have similar _structural_ issues as the problems I noted for `SyncProfileService` in https://github.com/pupilfirst/pupilfirst/pull/1627#discussion_r1697375280